### PR TITLE
Make refactor

### PR DESCRIPTION
**Pull request by CrispusDH**
- refactor `pages/`
- change locators in `widgets`
- set `directTrue` and `waiForAngular(false)`

### CODE REVIEW

File: protractor_demo/pages/search.ts [hunk 1]
```diff
@@ -1,22 +1,26 @@
...
 export class Search {
+    private readonly navigation: Navigation;
+    private readonly input: ElementFinder;
 
-    private readonly navigationMenu = new Navigation();
-    private readonly searchInput = browser.$('[name="q"]');
+    constructor() {
+        this.navigation = new Navigation();
+        this.input = $('[name="q"]');
+    }
```
Comment (AleksanderPopov) [inline protractor_demo/pages/search.ts]: same about constructor

File: protractor_demo/pages/search.ts [hunk 2]
```diff
@@ -1,22 +1,26 @@
...
 
-    async open() {
-        await browser.get('http://google.com/ncr');
+    public async open(url: string = 'http://google.com/ncr'): Promise<void> {
```
Comment (AleksanderPopov) [inline protractor_demo/pages/search.ts]: url for google search page is static - you dont need to pass anything in method which should open page (url especially)

File: protractor_demo/pages/search.ts [hunk 3]
```diff
@@ -1,22 +1,26 @@
...
+        await browser.get(url);
     }
 
-    async search(text) {
-        await browser.wait(ExpectedConditions.visibilityOf(this.searchInput), 5000);
-        await this.searchInput.sendKeys(text);
+    public async search(text: string): Promise<void> {
+        await browser.wait(ExpectedConditions.elementToBeClickable(this.input), 5000);
```
Comment (AleksanderPopov) [inline protractor_demo/pages/search.ts]: if wait for visibility is enough and you dont want to click - leave visibility condition :)

File: protractor_demo/pages/translate.ts [hunk 4]
```diff
@@ -1,39 +1,54 @@
...
 
 export class Translate {
+    private readonly source: ElementFinder;
+    private readonly translation: ElementFinder;
+    private readonly openSource: ElementFinder;
+    private readonly targetSource: ElementFinder;
+    private readonly slSearch: ElementFinder;
+    private readonly tlSearch: ElementFinder;
+
+    constructor() {
```
Comment (AleksanderPopov) [inline protractor_demo/pages/translate.ts]: same about constructor (here it cost 9 extra lines in file which have ~50 lines :) )

File: protractor_demo/pages/translate.ts [hunk 5]
```diff
@@ -1,39 +1,54 @@
...
+        this.slSearch = $('#sl_list-search-box');
+        this.tlSearch = $('#tl_list-search-box');
+    }
+
 
-    async translate(text: string) {
-        await browser.$('#source').sendKeys(text + protractor.Key.ENTER);
+    public async translate(text: string): Promise<void> {
+        await browser.wait(ExpectedConditions.elementToBeClickable(this.source), 3000);
+        await this.source.click();
```
Comment (AleksanderPopov) [inline protractor_demo/pages/translate.ts]: why click here?

File: protractor_demo/pages/translate.ts [hunk 6]
```diff
@@ -1,39 +1,54 @@
...
+        await this.source.sendKeys(text + protractor.Key.ENTER);
     }
 
-    async shouldHaveTranslation(translation: string) {
-        await browser.wait(async function () {
-            browser.$('span.translation').getText().then(text => text.includes(translation));
+    public async shouldHaveTranslation(translation: string): Promise<void> {
+        await browser.wait(async () => {
+            await browser.wait(ExpectedConditions.elementToBeClickable(this.translation), 3000);
```
Comment (AleksanderPopov) [inline protractor_demo/pages/translate.ts]: wait inside of wait? why just dont check visibility explicitly, without inner waits?

File: protractor_demo/pages/translate.ts [hunk 7]
```diff
@@ -1,39 +1,54 @@
...
-        await browser.wait(ExpectedConditions.visibilityOf(button), 3000);
-        await button.click();
-
-        const searchBox = browser.$('.tlid-open-source-language-list');
-        await browser.wait(ExpectedConditions.visibilityOf(searchBox), 3000);
-        await browser.$('#sl_list-search-box').sendKeys(lang + protractor.Key.ENTER);
+    private async chooseFromLang(lang: string): Promise<void> {
+        await browser.wait(ExpectedConditions.elementToBeClickable(this.openSource), 3000);
+        await this.openSource.click();
+        await browser.wait(ExpectedConditions.elementToBeClickable(this.slSearch), 3000);
```
Comment (AleksanderPopov) [inline protractor_demo/pages/translate.ts]: again, waiting for clickable and dont click

File: protractor_demo/widgets/navigation.ts [hunk 8]
```diff
@@ -1,28 +1,30 @@
...
+        const menu = await this.findMenu('Translate');
+        await browser.wait(ExpectedConditions.elementToBeClickable(menu), 3000);
+        await menu.click();
+    }
 
-        const webelements = await this.menus.getWebElements();
-        for (const webelement of webelements) {
-            if ((await webelement.isDisplayed()) && ('Translate' === await webelement.getText())) {
-                await webelement.click();
+    private async findMenu(name: string): Promise<ElementFinder> {
```
Comment (AleksanderPopov) [inline protractor_demo/widgets/navigation.ts]: dont declare new methods :)

File: protractor_demo/pages/results.ts [hunk 9]
```diff
@@ -1,27 +1,31 @@
...
-    async followNthResultLink(index) {
-        await browser.wait(async function () {
-            const webelements = await browser.$$('.g .r').getWebElements();
-            const resultContainer = await webelements[index].findElement(by.css('a'));
-            return resultContainer.isDisplayed();
-        }, 3000);
-        const webelements = await browser.$$('.g .r').getWebElements();
-        await webelements[0].findElement(by.css('a')).then(webelement => webelement.click());
+    public async followNthResultLink(index: number): Promise<void> {
+        const container: ElementFinder = await this.elements.get(index);
```
[the remark this answers is not in the document]
Comment (CrispusDH) [reply protractor_demo/pages/results.ts]: I use it on the next line, in such way I split some logical parts

File: protractor_demo/pages/translate.ts (hunk 5 above)
Comment (CrispusDH) [reply protractor_demo/pages/translate.ts]: it redundant here, good catch

File: protractor_demo/pages/translate.ts (hunk 6 above)
Comment (CrispusDH) [reply protractor_demo/pages/translate.ts]: I will split it

**Pull request merged**